Code/Framework/AzToolsFramework/AzToolsFramework/Viewport/ViewportMessages.cpp

```diff
@@ -66,15 +67,19 @@ namespace AzToolsFramework
 
     AZ::Vector3 FindClosestPickIntersection(const AzFramework::RenderGeometry::RayRequest& rayRequest, const float defaultDistance)
     {
-        AzFramework::RenderGeometry::RayResult renderGeometryIntersectionResult;
+        // attempt a ray intersection with any visible mesh or terrain and return the intersection position if successful
+        AZ::EBusReduceResult<AzFramework::RenderGeometry::RayResult, AzFramework::RenderGeometry::RayResultClosestAggregator> renderGeometryIntersectionResult;
         AzFramework::RenderGeometry::IntersectorBus::EventResult(
             renderGeometryIntersectionResult, AzToolsFramework::GetEntityContextId(),
             &AzFramework::RenderGeometry::IntersectorBus::Events::RayIntersect, rayRequest);
+        AzFramework::Terrain::TerrainDataRequestBus::BroadcastResult(
+            renderGeometryIntersectionResult,
+            &AzFramework::Terrain::TerrainDataRequests::GetClosestIntersection,
+            rayRequest);

```

Comment:
That's exactly why I added virtual EntityContextId GetTerrainRaycastEntityContextId() const = 0; which you asked about above - the TerrainRaycastContext implements AzFramework::RenderGeometry::IntersectorBus::Handler, but it has to connect to that bus using an AzFramework::EntityContextId. So we can easily change this to:

```
        const AzFramework::EntityContextId terrainEntityContextId = ???;
        AzFramework::RenderGeometry::IntersectorBus::EventResult(
            renderGeometryIntersectionResult, AzToolsFramework::GetEntityContextId(),
            &AzFramework::RenderGeometry::IntersectorBus::Events::RayIntersect, rayRequest);
```

but the only way to get terrainEntityContextId  is via the TerrainDataRequestBus, so I figured why not save an EBus call and just call AzFramework::Terrain::TerrainDataRequests::GetClosestIntersection directly.

I don't care either way, and if there's a better place for the terrainEntityContextId to live I'm fine moving it.

Reply:
I might be misunderstanding how this works. The entity context id should either be the tools context or the game context, right? The terrain shouldn't have its own unique one I wouldn't think. I would've thought that the terrain raycast would listen to whichever entity context is currently active, so the one existing IntersectorBus call would route to the terrain raycast. Would that work, or am I missing something?

Reply:
I'm not sure of the intent behind the design of the IntersectorBus / IntersectorInterface, but it's addressable using an EntityContextId (so anything that connects to it has to provide one), and it uses a handler policy of EBusHandlerPolicy::Single so there can only be one handler for each id connected to the bus.

Both the `EditorIntersectorComponent` and `GameIntersectorComponent` create instances of the `RenderGeometry::Intersector` class, which implements/connects to the `RenderGeometry::IntersectorBus` interface using the EditorEntityContextId and GameEntityContextId respectively.

The `RenderGeometry::Intersector` class is an:
```
        //! Implementation of IntersectorBus interface, this class contains a cached AABB list
        //! of the connected entities to the intersection bus and calculates performant ray intersections against them
```

So the only way to have terrain respond to calls made over the `RenderGeometry::IntersectorBus` is to:
a. Implement it directly using a new terrain specific EntityContextId, which is the route I took because terrain doesn't fall exclusively under either of the two existing EntityContextIds (EditorEntityContextId or GameEntityContextId).
b. Make it so both the `EditorIntersectorComponent` and `GameIntersectorComponent` have their `RenderGeometry::Intersector` instances account for terrain, which doesn't seem to make sense at the moment because the TerrainSystem is not an Entity and `RenderGeometry::Intersector` only appears to work with things that have an AZ::EntityId

I could very well be missing something, and I really don't mind how we go about it, but this is the only thing that made sense to me at the moment given how everything is currently setup.

Reply:
You should be able to use something like this...

```
static AzFramework::EntityContextId FindOwningContextId(const AZ::EntityId entityId)
{
    AzFramework::EntityContextId contextId = AzFramework::EntityContextId::CreateNull();
    AzFramework::EntityIdContextQueryBus::EventResult(
        contextId, entityId, &AzFramework::EntityIdContextQueries::GetOwningContextId);
    return contextId;
}
```

to get the EntityContextId from the Entity (there's one for Edit time and one for Game time)

See this PR as an example - https://github.com/o3de/o3de/pull/6452

Reply:
The TerrainSystem is not an entity though, and even if we end up getting the correct EntityContextId (Game vs Editor) somehow, I don't think that helps because the TerrainSystem won't be able to connect to the `RenderGeometry::IntersectorBus` using either of them, because the `EditorIntersectorComponent` and `GameIntersectorComponent` are already using those EntityContextIds to connect to that EBus.

Can we change the the IntersectorBus / IntersectorInterface to use a a handler policy of EBusHandlerPolicy::Multiple?

Reply:
Hmm I think having multiple things connected should be okay 🤔 Sorry I haven't had a chance to look over the change in full yet, it might be you need to use `IntersectionRequestBus` in `GeometryIntersectionBus.h` which is 'Id'ed off the `EntityContextId` and `EntityId` pair. If there's no entity involved this might make it more complicated though, I'll take another look tomorrow. The owner of that system has moved on though so if there are updates/improvements to make it more flexible for this use-case please have at it 🙂

Reply:
No worries! As I mentioned earlier I could very well be missing something, and I really don't mind how we go about it, but this seemed like the only way to get it working (without expanding the scope of the changes) given how everything is currently setup. It works for the current use case of creating entities on top of terrain, but I'm happy to make wider reaching changes if it's appropriate.

Reply:
I think we'll need to leave this for a follow-up PR to see if it can be improved. Since the above code works, and there isn't an immediately obvious way to improve it with the existing intersector bus designs, it seems reasonable to submit as-is and then refactor if we can come up with something better.